[NEMO-64] Fix map stage hang under DataSkewPolicy

jeongyooneo opened the pull request:
JIRA: [NEMO-64: Fix map stage hang under DataSkewPolicy](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-64),
[NEMO-181: Fix DataSkewPolicy bug for multiple shuffles](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-181)


**Major changes:**
- Data for dynamic optimization is aggregated at `DynOptDataHandler` in Scheduler instead of a variable in `MetricCollectionBarrierVertex`. Updating `MetricCollectionBarrierVertex` as data arrives in RuntimeMaster clashed with serializing IRDAG for scheduled Tasks, which caused `ConcurrentModificationException` and silently killed the Scheduler
- Identifies the target edge to optimize via MetricCollectionProperty in case multiple shuffles are involved.

**Minor changes to note:**
- Removed now unused entries in `dataSizeMetricMsg`
- Added `DataSkewMetricProperty` and `MetricFactory` to make task hash ranges of shuffle edges as an execution property, so that RuntimePass can optimize the given IR DAG

**Tests for the changes:**
- N/A(No new features, covered by `PerKeyMedianITCase`)

**Other comments:**
- Data for dynamic optimization will be aggregated in designated vertex and not in the RuntimeMaster via upcoming PR for [NEMO-99](https://issues.apache.org/jira/browse/NEMO-99)

resolves [NEMO-64](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-64), [NEMO-181](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-181)

johnyangk commented on common/src/main/java/edu/snu/nemo/common/ir/vertex/IRVertex.java:
```diff
@@ -30,13 +30,15 @@
  */
 public abstract class IRVertex extends Vertex {
   private final ExecutionPropertyMap<VertexExecutionProperty> executionProperties;
+  private boolean stagePartitioned;
```
I'd remove this, as `IRVertex` should not know about stages.

johnyangk commented on compiler/optimizer/src/main/java/edu/snu/nemo/compiler/optimizer/pass/compiletime/annotating/SkewResourceSkewedDataPass.java:
```diff
@@ -53,10 +53,16 @@ private boolean hasMetricCollectionBarrierVertexAsParent(final DAG<IRVertex, IRE
         .filter(v -> hasMetricCollectionBarrierVertexAsParent(dag, v)
             && !v.getExecutionProperties().containsKey(ResourceSkewedDataProperty.class))
-        .forEach(v -> v.getExecutionProperties().put(ResourceSkewedDataProperty.of(true)));
+        .forEach(childV -> {
```
childV -> childVertex

johnyangk commented on runtime/common/src/main/java/edu/snu/nemo/runtime/common/optimizer/RunTimeOptimizer.java:
```diff
@@ -36,24 +42,25 @@ private RunTimeOptimizer() {
+      // Metric data for DataSkewRuntimePass is a map of <hashrange, partition size>.
+      final Map<Integer, Long> metricData = (Map<Integer, Long>) metric;
+      final DAG<IRVertex, IREdge> newIrDAG =
+          new DataSkewRuntimePass().apply(originalPlan.getIrDAG(), metricData);
```
add a TODO here?
`DataSkewRuntimePass` is hard-coded.

johnyangk commented on runtime/executor/src/test/java/edu/snu/nemo/runtime/executor/datatransfer/DataTransferTest.java:
```diff
@@ -300,6 +303,15 @@ private void writeAndRead(final BlockManagerWorker sender,
     dummyIREdge.setProperty(DataPersistenceProperty.of(DataPersistenceProperty.Value.Keep));
     dummyIREdge.setProperty(EncoderProperty.of(ENCODER_FACTORY));
     dummyIREdge.setProperty(DecoderProperty.of(DECODER_FACTORY));
+    if (dummyIREdge.getPropertyValue(CommunicationPatternProperty.class).get()
```
I'm not sure the `DataSkewMetricProperty` set in L306-L314, L398-L405 is used/checked afterwards. A standalone test may be a better option.

jeongyooneo replied:
It is actually used in `InputReader#read`(L355 and others). As now input data for each task is set as Execution Properties(`DefaultMetricPass`) instead of the constructor in `StageEdge`, we need to initialize it here for shuffle read.

johnyangk commented on runtime/master/src/main/java/edu/snu/nemo/runtime/master/DynOptDataHandler.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Handler for aggregating data used in dynamic optimization.
+ */
+public interface DynOptDataHandler {
```
Provide updateDynOptData and getDynOptData as interface methods?

johnyangk commented on runtime/master/src/main/java/edu/snu/nemo/runtime/master/resource/ExecutorRepresenter.java:
```diff
@@ -107,19 +110,26 @@ public void onTaskScheduled(final Task task) {
     runningTaskToAttempt.put(task, task.getAttemptIdx());
     failedTasks.remove(task);
-
+    final byte[] serialized;
```
Revert L113-L120 for multi-threaded serialization?

johnyangk commented on runtime/master/src/main/java/edu/snu/nemo/runtime/master/scheduler/BatchScheduler.java:
```diff
@@ -93,6 +93,7 @@ private BatchScheduler(final TaskDispatcher taskDispatcher,
           .subscribe(updatePhysicalPlanEventHandler.getEventClass(), updatePhysicalPlanEventHandler);
     }
     this.executorRegistry = executorRegistry;
+    this.dataSkewDynOptDataHandler = new DataSkewDynOptDataHandler();
```
Do we need more than one handler to handle multiple RuntimePasses in the future?
Add TODO here?

jeongyooneo replied:
For that case, I'll refactor `DynOptDataHandler` to keep a list of data handlers and add comments on it.

johnyangk commented on runtime/master/src/main/java/edu/snu/nemo/runtime/master/scheduler/BatchScheduler.java:
```diff
@@ -338,11 +337,11 @@ private void doSchedule() {
           taskId,
           attemptIdx,
-          stageToSchedule.getExecutionProperties(),
+          stageToSchedule.getExecutionProperties().getCopy(),
```
Revert L340-L344?

johnyangk commented on runtime/master/src/main/java/edu/snu/nemo/runtime/master/scheduler/TaskDispatcher.java:
```diff
@@ -130,8 +130,6 @@ private void doScheduleTaskList() {
           // update metadata first
           planStateManager.onTaskStateChanged(task.getTaskId(), TaskState.State.EXECUTING);
 
-          LOG.info("{} scheduled to {}", task.getTaskId(), selectedExecutor.getExecutorId());
```
Undo this?

johnyangk commented on runtime/common/src/main/proto/ControlMessage.proto:
```diff
@@ -117,8 +117,6 @@ message BlockStateChangedMsg {
 message DataSizeMetricMsg {
     // TODO #96: Modularize DataSkewPolicy to use MetricVertex and BarrierVertex.
     repeated PartitionSizeEntry partitionSize = 1;
-    required string blockId = 2;
```
id of the metric vertex?
The runtime master needs to know which metric vertex is this proto for?

jeongyooneo replied:
Now metric vertex only acts as a barrier, and all run-time optimization is done via ExecutionProperties of IREdges. So id of metric vertex isn't needed anymore.

johnyangk commented on runtime/common/src/main/java/edu/snu/nemo/runtime/common/plan/Task.java:
```diff
@@ -67,6 +65,17 @@ public Task(final String planId,
     this.irVertexIdToReadable = irVertexIdToReadable;
   }
 
+  public Task clone() {
```
We may no longer need this?

johnyangk commented:
@seojangho Please also take a look at this PR. 😄

johnyangk commented:
It'd be also nice to add a unit test that hangs with the current master branch code, and does not hang with this PR.

jeongyooneo commented:
@johnyangk Thanks for the review, I've addressed most of your comments.
As there are upcoming PRs for refactoring dynamic optimization which change current execution model of dynamic optimization. Can I resolve the left issues, such as adding the suggested unit test, with those PRs?
